doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,176 @@
+
+    Informally, updates at times not greater than the `since` frontier may have been consolidated with other updates.
+    They are not discarded, but their times may have been advanced to the `since` frontier to allow cancelation and consolidation of updates to occur.
+    This is analogous to "version vacuuming" in multiversioned databases.
+    As the system operates, the `since` frontier may advance to allow compaction of data.
+    This happens subject to constraints, and only when correctness permits.
+
+    We think of `since` as the "read frontier": times greater or equal to it may still be correctly read from the TVC.
+
+*   The `upper` frontier, which communicates that for times `time` not greater or equal to `upper`, the updates at `time` are complete.
+
+    The upper frontier provides the information about how a TVC may continue to evolve.
+    All future updates will necessarily have their `time` greater or equal to the `upper` frontier.
+    As the system operates, we expect the `upper` frontiers to advance indicating that the contents of the collection are known for more times.
+    This happens only when we become certain that updates for a time are complete.
+
+    We think of `upper` as the "write frontier": times greater or equal to it may still be written to the TVC.
+
+These two frontiers mean that for any `time` greater or equal to `since`, but not greater or equal to `upper`, we can exactly reconstruct the collection at `time`.
+If `time` is not greater or equal to `since`, we may have permanently lost historical distinctions that allow us to know the contents at `time`.
+If `time` is greater or equal to `upper`, we may still receive changes to the collection that occur at `time`.
+The two frontiers never move backward, and often move forward as the system operates.
+
+At each moment, Materialize's representation of a TVC is as a set of update triples `(data, time, diff)` and the two frontiers `since` and `upper`.
+We call this pair of set of updates and `(since, upper)` frontiers a *partial TVC* or pTVC.
+
+## Evolution
+
+Materialize maintains a (partial) map from `GlobalId`s to pTVCs.
+A `GlobalId` is never re-used to mean a different TVC.
+
+The life cycle of each `GlobalId` follows several steps:
+
+0. The `GlobalId` is initially not yet bound.
+
+    In this state, the `GlobalId` cannot be used, as we do not yet know what its initial `since` will be.
+
+1. The `GlobalId` is bound to some definition, with an initially equal `since` and `upper`.
+
+    This means that the collection is initially not useful, as it cannot be queried at any time greater or equal to `since` but not greater or equal to `upper`.
+    However, the initial value of `since` indicates that one should expect the data to eventually become available for each time greater or equal to `since`.
+    The partial TVC can now be used in queries, though results may not be immediately available.
```

Comment:
This is a bit confusing, as the statement contradicts a statement in the same paragraph and under the same point.
Maybe this should be moved as a special case of a.

Also, if I start with `since_1 = upper_1` and advance both to `since_2 = upper_2`, is the `GlobalId` unbound again?

Reply:
Can you say more about which statement this contradicts? Do you mean the "as it cannot be queried .." above? That should be an easy fix by replacing "queried" there with "observed".

This is definitely not a special case, as it is important that there is a transition from "absent" to "present, with `since`" to avoid misleading the user into thinking they can correctly read data at times not greater or equal to the initial `since`.

The `GlobalId` is not unbound once it is bound. It ends up "unusable" but not unbound.

Reply:
> Do you mean the "as it cannot be queried .." above?
Yes, precisely. 

There might be a distinction between a `GlobalId` being "useful" and being "used in queries" which I am not catching.

I understand that `since = upper` implies that the set of times `t` for which you can observe the associated TVC is empty, and I think this is what you mean by "not useful".

What I am not sure is whether this means that

1. one can reference the `GlobalId` in queries only after `upper` advances, or
2. this is possible immediately after `GlobalId` is bound, but the fact that TVC is not useful initially manifests in empty results.

My intuition is (1), but the way the second sentence is written makes it a bit unclear. If this is the case, maybe replace the third sentence as follows:

> Once the `upper` frontier advances, the partial TVC can be used in queries.

Reply:
I think I have a bunch of loose text like this that I should edit, where "used in queries" is sometimes technical and important, and sometimes casual. I'll scan for all of this and edit appropriately.

The intent is that there is a specific moment, once an id is bound, that it is now correct to use the id in other commands. Whether those commands will responds promptly or not is a different issue. I'll try and cinch the language up, thanks!